libvast/src/evaluate.cpp

```diff
@@ -0,0 +1,465 @@
+  const auto num_rows = slice.rows();
+  const auto evaluate_predicate = detail::overload{
+    [](const auto&, relational_operator, const auto&, const ids&) -> ids {
+      die("predicates must be normalized and bound for evaluation");
+    },
+    [&](const meta_extractor& lhs, relational_operator op, const data& rhs,
+        ids selection) -> ids {
+      // If no bit in the selection is set we have no results, but we can avoid
+      // an allocation by simply returning the already empty selection.
+      if (!any(selection))
+        return selection;
+      if (evaluate_meta_extractor(slice, lhs, op, rhs))
+        return selection;
+      return ids{offset + num_rows, false};
+    },
+    [&](const data_extractor& lhs, relational_operator op, const data& rhs,
+        const ids& selection) -> ids {
+      if (!any(selection))
+        return ids{offset + num_rows, false};
+      const auto index
+        = caf::get<record_type>(slice.layout()).resolve_flat_index(lhs.column);
+      const auto type = caf::get<record_type>(slice.layout()).field(index).type;
+      const auto array = static_cast<arrow::FieldPath>(index)
+                           .Get(*to_record_batch(slice))
+                           .ValueOrDie();
+      VAST_ASSERT(array);
+      switch (op) {
+#define VAST_EVAL_DISPATCH(op)                                                 \
+  case relational_operator::op: {                                              \
+    auto f = [&]<concrete_type Type, class Rhs>(                               \
+               Type type, const Rhs& rhs) noexcept -> ids {                    \
+      return column_evaluator<relational_operator::op, Type,                   \
+                              view<Rhs>>::evaluate(type, offset, *array,       \
+                                                   make_view(rhs), selection); \
+    };                                                                         \
+    return caf::visit(f, type, rhs);                                           \
+  }
+        VAST_EVAL_DISPATCH(equal);
+        VAST_EVAL_DISPATCH(not_equal);
+        VAST_EVAL_DISPATCH(in);
+        VAST_EVAL_DISPATCH(less);
+        VAST_EVAL_DISPATCH(not_in);
+        VAST_EVAL_DISPATCH(match);
+        VAST_EVAL_DISPATCH(not_match);
+        VAST_EVAL_DISPATCH(greater);
+        VAST_EVAL_DISPATCH(greater_equal);
+        VAST_EVAL_DISPATCH(less_equal);
+        VAST_EVAL_DISPATCH(ni);
+        VAST_EVAL_DISPATCH(not_ni);
+#undef VAST_EVAL_DISPATCH
+      }
+      die("unreachable");
+    },
+  };
+  const auto evaluate_expression
+    = [&](const auto& self, const expression& expr, ids selection) -> ids {
+    const auto evaluate_expression_impl = detail::overload{
+      [&](const caf::none_t&, const ids&) {
+        return ids{offset + num_rows, false};
+      },
+      [&](const negation& negation, const ids& selection) {
+        return selection ^ self(self, negation.expr(), selection);
```

Comment:
Can you document why XOR does The Right Thing?

Reply:
I can, but note that this code was only moved, we did not touch it at all.

Reply:
I added a comment for this, please check back. :)

Reply:
I'm not quite getting it. This lacks context for me:
> For bits that are zero because of the offset the result bitmap is also always guaranteed to be zero [..]

Reply:
Is it clearer to you if I write that for `selection & self(self, negation.expr(), selection)` is the same as `self(self, negation.expr(), selection)`?

Reply:
No. 🙂

It should be the mental model that's conveyed, i.e., why the logical operation on bitmaps make sense.

But we don't have to go crazy here. This suits the bill. Someone who goes deeper can draw it out on paper.